### PR TITLE
[CLEANUP] Refactor `parseCssRules`: parse all CSS

### PR DESCRIPTION
**Pull request by JakeQZ**
Refactored `parseCssRules` so that it can now parse an entire CSS string,
including @media rules.  It now returns two arrays, in an array with the
following keys:
- "inlineable" => The inlineable rules that would have been previously returned;
- "uninlineable" => The uninlineable rules, currently those from within @media
  rules.

The original line number information from the CSS is available in both arrays,
so the original ordering of all the rules (that will be required for #280) can
be determined.

A temporary method `combineCssMediaRules` has been added to recombine the @media
rules back into a string of CSS as expected by `copyCssWithMediaToStyleNode`,
pending refactoring of that method to instead accept a parsed rule array as its
parameter.  This does mean that, for now, the rules within each @media rule are
parsed again.

### CODE REVIEW

**Comment by JakeQZ** [comment]
As you'll see, the code in `combineCssMediaRules` is somewhat messy and I think would benefit from a `CssConcatenator` class to abstract the functionality for combining rules with the same media query, selectors or declarations block.  (This functionality will still be required after refactoring `copyCssWithMediaToStyleNode`, then by the refactored method itself.)  This I propose to add in the next incremental PR, though it could be included in this PR if desired...

File: src/Emogrifier.php [hunk 1]
```diff
@@ -670,42 +675,60 @@ private function parseCssRules($css)
...
-                    $hasSupportedPseudoClass = (bool)preg_match(
-                        '/:(\\S+\\-(child|type\\()|not\\([[:ascii:]]*\\))/i',
-                        $selector
-                    );
-                    if ($hasPseudoElement || ($hasAnyPseudoClass && !$hasSupportedPseudoClass)) {
+            $splitCss = $this->splitCssAndMediaQuery($css);
+
+            $cssRules = [
+                'inlineable' => [],
+                'uninlineable' => []
```
Comment (oliverklee) [inline src/Emogrifier.php]: Please add a trailing comma also after the last element of a multiline array. This makes appending elements possible without having to modify the existing line.

File: src/Emogrifier.php [hunk 2]
```diff
@@ -670,42 +675,60 @@ private function parseCssRules($css)
...
 
-                    $cssRules[] = [
-                        'selector' => trim($selector),
-                        'declarationsBlock' => $cssDeclaration,
-                        // keep track of where it appears in the file, since order is important
-                        'line' => $key,
-                    ];
+                    $selectors = explode(',', $cssRule[1]);
+                    foreach ($selectors as $selector) {
+                        // don't process pseudo-elements and behavioral (dynamic) pseudo-classes;
```
Comment (oliverklee) [inline src/Emogrifier.php]: Can we have this part (within the foreach loop of $selector, or most of it) in a separate private method?

File: src/Emogrifier.php [hunk 3]
```diff
@@ -670,42 +675,60 @@ private function parseCssRules($css)
...
-                    );
-                    if ($hasPseudoElement || ($hasAnyPseudoClass && !$hasSupportedPseudoClass)) {
+            $splitCss = $this->splitCssAndMediaQuery($css);
+
+            $cssRules = [
+                'inlineable' => [],
+                'uninlineable' => []
+            ];
+            // keep track of where each rule appears in the file, since order is important
+            $key = 0;
```
Comment (oliverklee) [inline src/Emogrifier.php]: Can't we have this as the key in the corresponding loop?

File: src/Emogrifier.php [hunk 4]
```diff
@@ -1069,6 +1092,78 @@ private function attributeValueIsImportant($attributeValue)
...
 
+    /**
+     * Reconstructs the CSS containing only allowed @media rules from the "uninlineable" CSS rules array returned by
+     * `parseCssRules`.
+     *
+     * @param string[][] $cssRules
+     *
+     * @return string
+     */
+    private function combineCssMediaRules(array $cssRules)
```
Comment (oliverklee) [inline src/Emogrifier.php]: Can you also please split up this method? It's far too long.

**Comment by oliverklee** [comment]
As for the `CssConcatenator` class, we will only be able to have this for the `CssInliner` class, not for `Emogrifier` anymore (as the `Emogrifier` class is Composer-optional, and the class will go away for version 4.0.0).

**Comment by oliverklee** [comment]
Oops, wrong button. Sorry, Reopening.

**Comment by JakeQZ** [comment]
> As for the CssConcatenator class, we will only be able to have this for the CssInliner class, not for Emogrifier anymore (as the Emogrifier class is Composer-optional, and the class will go away for version 4.0.0).

Would something like 
```
        // support use without autoload
        if (!class_exists('Pelago\\Emogrifier\\CssConcatenator')) {
            require_once __DIR__ . '/Emogrifier/CssConcatenator.php';
        }
```
not be possible within the Emogrifier class?  I understand the need to support those not using autoload, but is it also a requirement to support grabbing just the `.php` file rather than the `src` directory (tree)?

File: src/Emogrifier.php (hunk 3 above)
Comment (JakeQZ) [reply src/Emogrifier.php]: Currently it needs to increment with the inner loop and carry on incrementing with each iteration of the outer loop.  However, it should be possible to have one double-loop doing the regex matching and `array_merge`-ing the results, then a second single loop processing the matches...

**Comment by oliverklee** [comment]
No, that would be okay for me.

File: src/Emogrifier.php (hunk 4 above)
Comment (JakeQZ) [reply src/Emogrifier.php]: This is why I'd like to add a `CssConcatenator` class.

**Comment by JakeQZ** [comment]
I've added a `CssConcatenator` class as a separate PR (#530).  I'll come back to this one when that is agreed.

File: src/Emogrifier.php (hunk 1 above)
Comment (JakeQZ) [reply src/Emogrifier.php]: Done.

File: src/Emogrifier.php (hunk 2 above)
Comment (JakeQZ) [reply src/Emogrifier.php]: I've moved the outer loop and regex matching into a separate method, so the loops are now very similar to how they were prior to this PR.

File: src/Emogrifier.php (hunk 3 above)
Comment (JakeQZ) [reply src/Emogrifier.php]: Changed as per comment above.

File: src/Emogrifier.php (hunk 4 above)
Comment (JakeQZ) [reply src/Emogrifier.php]: Now using `CssConcatenator`.

**Comment by JakeQZ** [comment]
Now made the changes suggested in review.  Also rebased and reverted changes to `CssInliner.php` which can be merged as a separate PR once agreed.

File: src/Emogrifier.php [hunk 5]
```diff
@@ -694,23 +701,62 @@ private function parseCssRules($css)
...
+            usort($cssRules['inlineable'], [$this, 'sortBySelectorPrecedence']);
 
             $this->caches[static::CACHE_KEY_CSS][$cssKey] = $cssRules;
         }
 
         return $this->caches[static::CACHE_KEY_CSS][$cssKey];
     }
 
+    /**
+     * Parse a string of CSS into the media query, selectors and declarations for each ruleset in order
```
Comment (oliverklee) [inline src/Emogrifier.php]: Third person, please ("Parses") and a period at the end of sentence.

File: src/Emogrifier.php [hunk 6]
```diff
@@ -694,23 +701,62 @@ private function parseCssRules($css)
...
+                        'media' => $cssRule['media'],
                         'selector' => trim($selector),
                         'declarationsBlock' => $cssDeclaration,
                         // keep track of where it appears in the file, since order is important
                         'line' => $key,
                     ];
+                    if ($cssRule['media'] === '') {
+                        $cssRules['inlineable'][] = $parsedCssRule;
+                    } else {
+                        $cssRules['uninlineable'][] = $parsedCssRule;
```
Comment (oliverklee) [inline src/Emogrifier.php]: I think we can shorten this to something like this:

```php
$ruleType = ($cssRule['media'] === '') ? 'inlineable' : 'uninlineable';
$cssRules[$ruleType] = $parsedCssRule;
```
